File: typescript/src/parallel-invocation.ts
import {
	type DurableContext,
	withDurableExecution,
} from "@aws/durable-execution-sdk-js";
import { MODEL } from "./utils/constants.js";
import { converse } from "./utils/converse.js";

const PROMPTS = [
	"Explain the benefits of",
	"Describe the challenges of",
	"Summarize the future of",
];

/**
 * Parallel Invocation Example
 *
 * Demonstrates invoking multiple prompts at once in a workflow,
 * for either breakdown of work or requesting multiple views/approaches on a single topic.
 */
export const handler = withDurableExecution(
	async (event: { topic?: string }, context: DurableContext) => {
		const topic = event.topic ?? "artificial intelligence";

		const result = await context.map(
			"Get perspectives",
			PROMPTS,
			async (_, prompt) => ({
				prompt,
				response: await converse(MODEL.SMALL, `${prompt} ${topic}`),
			}),
			{ itemNamer: (_, i) => `prompt-${i}` },
		);

		return {
			topic,
			perspectives: result.getResults(),
		};
	},
);

File: python/src/parallel_invocation.py
from aws_durable_execution_sdk_python import DurableContext, durable_execution

from src.utils.constants import MODEL
from src.utils.converse import converse

PROMPTS = [
    "Explain the benefits of",
    "Describe the challenges of",
    "Summarize the future of",
]


@durable_execution
def handler(event: dict, context: DurableContext):
    event = event or {}
    topic = event.get("topic", "artificial intelligence")

    result = context.map(
        PROMPTS,
        lambda ctx, prompt, idx, items: {
            "prompt": prompt,
            "response": converse(MODEL["SMALL"], f"{prompt} {topic}"),
        },
        "Get perspectives",
    )

    return {"topic": topic, "perspectives": result.get_results()}
